src/tools/pkcs11-tool.c

```diff
@@ -2023,16 +2036,24 @@ do_read_key(unsigned char *data, size_t data_len, int private, EVP_PKEY **key)
 
 #define RSA_GET_BN(RSA, LOCALNAME, BNVALUE) \
 	do { \
-		RSA->LOCALNAME = malloc(BN_num_bytes(BNVALUE)); \
-		if (!RSA->LOCALNAME) \
-			util_fatal("malloc() failure"); \
-		RSA->LOCALNAME##_len = BN_bn2bin(BNVALUE, RSA->LOCALNAME); \
+		if (BNVALUE) { \
+			RSA->LOCALNAME = malloc(BN_num_bytes(BNVALUE)); \
```

Comment:
In parse_rsa_pkey one of the parameters is struct rsakey_info *rsa. 
The macro RSA_GET_BN(RSA, LOCALNAME, BNVALUE) is only used in parse_pkey. 
So the macro is operating on a struct rsakey_info not an OpenSSL RSA object. 

The new code in parse_rsa_pkey uses the RSA_get0_* routines to get pointers to:
 +	const BIGNUM *r_n, *r_e, *r_d;
+	const BIGNUM *r_p, *r_q;
+	const BIGNUM *r_dmp1, *r_dmq1, *r_iqmp;

then the RSA_GET_BN macro is use to extract from teh BIGNUMs what it needs.